src/container/srv_target.c

```diff
@@ -181,65 +365,87 @@ cont_stop_dtx_reindex_ult(struct ds_cont_child *cont)
 	cont->sc_dtx_reindex_abort = 0;
 }
 
-static int
-cont_start_agg_ult(struct ds_cont_child *cont)
-{
-	int rc;
-
-	D_ASSERT(cont != NULL);
-	if (cont->sc_abort_vos_aggregating || cont->sc_vos_aggregating)
-		return 0;
+/* ds_cont_child *******************************************************/
 
-	cont->sc_vos_aggregating = 1;
-	ds_cont_child_get(cont);
-	rc = dss_ult_create(ds_cont_aggregate_ult, cont,
-			    DSS_ULT_AGGREGATE, DSS_TGT_SELF, 0, NULL);
-	if (rc) {
-		D_ERROR(DF_UUID": Failed to create aggregation ULT; rc %d\n",
-			DP_UUID(cont->sc_uuid), rc);
-		cont->sc_vos_aggregating = 0;
-		ds_cont_child_put(cont);
-	}
-	return rc;
+static inline struct ds_cont_child *
+cont_child_obj(struct daos_llink *llink)
+{
+	return container_of(llink, struct ds_cont_child, sc_list);
 }
 
-static void
-cont_stop_agg_ult(struct ds_cont_child *cont)
+static int
+cont_child_alloc_ref(void *co_uuid, unsigned int ksize, void *po_uuid,
+		     struct daos_llink **link)
 {
-	int rc;
+	struct ds_cont_child	*cont;
+	int			rc;
 
-	if (!cont->sc_vos_aggregating)
-		return;
+	D_ASSERT(po_uuid != NULL);
+	D_DEBUG(DF_DSMS, DF_CONT": opening\n", DP_CONT(po_uuid, co_uuid));
 
-	D_DEBUG(DF_DSMS, DF_CONT": Stopping aggregation ULT\n",
-		DP_CONT(NULL, cont->sc_uuid));
+	D_ALLOC_PTR(cont);
+	if (cont == NULL)
+		return -DER_NOMEM;
 
-	if (cont->sc_agg_ult) {
-		dss_ult_wakeup(cont->sc_agg_ult);
-		dss_sleep_ult_destroy(cont->sc_agg_ult);
-		cont->sc_agg_ult = NULL;
+	rc = ABT_mutex_create(&cont->sc_mutex);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
 	}
 
-	cont->sc_abort_vos_aggregating = 1;
-	rc = vos_cont_ctl(cont->sc_hdl, VOS_CO_CTL_ABORT_AGG);
-	if (rc)
-		D_ERROR(DF_UUID": Abort aggregation failed. %d\n",
-			DP_UUID(cont->sc_uuid), rc);
+	rc = ABT_cond_create(&cont->sc_dtx_resync_cond);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out_mutex;
+	}
 
-	while (cont->sc_vos_aggregating)
-		ABT_thread_yield();
+	cont->sc_pool = ds_pool_child_lookup(po_uuid);
+	if (cont->sc_pool == NULL) {
+		rc = -DER_NO_HDL;
+		goto out_cond;
+	}
+
+	rc = vos_cont_open(cont->sc_pool->spc_hdl, co_uuid, &cont->sc_hdl);
```

Comment:
I don't think vos_cont_open() should be tied with DAOS container open, even if the container isn't opened by any client, we still need vos container opened for background jobs like aggregation, etc.

Reply:
Hmm,  if the container is not being accessed, then it means no I/O on the container? Normally, this container does not need to aggregate at all?

But anyway, I do not understand why vos_cont_open can not be tied with DAOS container open? I mean even for aggregation, it can still explicitly open the vos container, instead of through ds_cont_lookup/create?